### PR TITLE
ICU-22511 Fix "vi" locale collator hang in comparison

### PR DESCRIPTION
**Pull request by FrankYFTang**
Avoid infinity loop by terminating and returning error if we loop the same for too many times.

##### Checklist

- [X] Required: Issue filed: https://unicode-org.atlassian.net/browse/ICU-22511
- [X] Required: The PR title must be prefixed with a JIRA Issue number. 
- [X] Required: The PR description must include the link to the Jira Issue, for example by completing the URL in the first checklist item
- [X] Required: Each commit message must be prefixed with a JIRA Issue number. 
- [x] Issue accepted (done by Technical Committee after discussion)
- [X] Tests included, if applicable
- [ ] API docs and/or User Guide docs changed or added, if applicable

### CODE REVIEW

**Comment by FrankYFTang** [comment]
@markusicu friendly ping. this bug was filed Nov 2, 2023 by fuzzer and and the PR is ready for review for a week.

**Comment by macchiati** [comment]
I think both are out of office for Thanksgiving week.

**Comment by macchiati** [comment]
I took a quick look at the code, and I am concerned. It appears to return as equal whenever three times in a row, a primary is equal. But then we would get the clearly incorrect:

aaah = aaagh

Each pass through the main loop should be resetting either the right or left CEs, or both. Because getting a ce advances an internal pointer, that should always terminate. So something fishy is going on.

I suggest that you augment your printout to provide more readable information, since lines like the following are too hard to follow:
ce from left 0xa4000500
ce from left 0x2a00000005009c00
...

I suggest when you print, break down the left and right ce values in to primary, secondary, tertiary, and put tabs between the fields  so that this can be loaded into a spreadsheet.
Eg something like:
...
Lp 2a0000 Ls 0500 Lt 0042
...

I can't remember, but you make also be able to fetch a code point offset from the ce; if so it would help to include that in the printout.

**Comment by FrankYFTang** [comment]
sorry, I assume you look at the Java code first. My java change is indeed wrong. I should throw there not return equal.

**Comment by macchiati** [comment]
Thanks. I'm still a bit worried about the change, because of the aaah case, so I'd appreciate a printout for the test case along the lines I suggested.

**Comment by richgillam** [comment]
I looked at this and didn't understand what it was doing or why.  I think I'll just defer to Mark here.

**Comment by FrankYFTang** [comment]
@markusicu ping

**Comment by FrankYFTang** [comment]
ping @markusicu . this is more urgent than other PR because it is now able to hang v8 with very simple script (see the bug for details)

**Comment by FrankYFTang** [comment]
Somehow the bug is fixed between 
https://github.com/unicode-org/icu/compare/788b89321454f3eeb20782032c8444aabba42417...699fb1dbc4cfbae6f78ff0b28570f44a20a7b149

per fuzzer. Obsolete this PR

**Comment by FrankYFTang** [comment]
reopen 
ok, the fuzzer is verify the fixed just because we increase the number of supported locale and change the testing locale to a different one. If we try the tests in this PR it is still broken.

**Comment by FrankYFTang** [comment]
The fix is not good enough.
